walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
 
 package com.google.sps;
 
+import com.google.sps.TimeRange;
+import java.lang.Math;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
 
+/**
+ * Determines potential meeting times based on a set of pre-existing events
+ * and mandatory/optional attendees who must be present.
+ */
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  /** Used to sort events by start time. */
+  public static final Comparator<Event> EVENT_COMPARATOR = new Comparator<Event>() {
+
+    @Override
+    public int compare(Event a, Event b) {
+      return TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen());
+    }
+  };
+
+  /** Finds potential meeting times */
+  public Collection<TimeRange> query(
+    Collection<Event> events,
+    MeetingRequest request
+  ) {
+    // Potential windows that only include mandatory attendees.
+    ArrayList<TimeRange> timeSlots = new ArrayList<>();
+    // Potential windows that include ALL mandatory + optional attendees.
+    ArrayList<TimeRange> timeSlotsWithOptional = new ArrayList<>();
+
+    ArrayList<Event> sortedEvents = new ArrayList<>(events);
+    sortedEvents.sort(EVENT_COMPARATOR);
+
+    /*
+     * We consider this problem in terms of 'windows' and 'blocks'.
+     * |-----A-----|                  |------A-----|
+     *       |----------B-----|       |---B---|
+     *
+     * |______________________||______|
+     *        block              window
+     *
+     * We iterate shifts to identify the right-most extent of a block.
+     * If we find a shift that start *after* this point, the difference is a window.
+     * That window is a possible meeting spot, if long enough.
+     */
+    int mandatoryBlockExtent = TimeRange.START_OF_DAY; // Based on people who *must* attend
+    int optionalBlockExtent = TimeRange.START_OF_DAY; // Based on people who *may* attend
+
+    // Whether any events exist for mandatory and optional attendees respectively.
+    boolean hasMandatoryEvents = false;
+    boolean hasOptionalEvents = false;
+
+    // Whether we should consider attendees of a given class.
+    // This is used to handle cases like all optional no mandatory, no attendees, etc.
+    boolean considerMandatoryAttendees = request.getAttendees().size() > 0;
+    boolean considerOptionalAttendees = request.getOptionalAttendees().size() > 0;
```

Comment:
Covering the case where there are no attendees would be easier if you just add a check to the top of the function.  The function is complex enough that you have to read the full code (or trust the tests) to convince yourself that the correct behaviour is performed in that edge case.